File: Algorithms/Easy/ReverseLinkedList.py
from typing import Optional

class ListNode:
    def __init__(self, val=0, next=None):
        self.val = val
        self.next = next

# Time complexity: O(n)
# Space complexity: O(1)
class Solution:
    def reverseList(self, head: Optional[ListNode]) -> Optional[ListNode]:
        reversed_head = None       

        while head:
            curr = ListNode(head.val)
            curr.next = reversed_head
            reversed_head = curr
            head = head.next

        return reversed_head

File: Algorithms/Easy/ReverseLinkedList.go
package easy

import (
	"github.com/parviz-yu/LeetCode/Algorithms/helpers"
)

// Time complexity: O(n)
// Space complexity: O(1)
func reverseList(head *helpers.ListNode) *helpers.ListNode {
	var prev *helpers.ListNode

	for head != nil {
		nxt := head.Next
		head.Next = prev
		prev = head
		head = nxt
	}

	return prev
}
